tools/checklocks/state.go

```diff
@@ -105,18 +155,24 @@ func (l *lockState) modify() {
 	}
 }
 
+func (l *lockState) getLock(s string) lockType {
+	if _, found := l.lockedMutexes[s]; !found {
+		l.lockedMutexes[s] = UNLOCKED
+	}
+
+	return l.lockedMutexes[s]
+}
+
 // isHeld indicates whether the field is held is not.
 func (l *lockState) isHeld(rv resolvedValue) (string, bool) {
 	if !rv.valid {
 		return rv.valueAsString(l), false
 	}
 	s := rv.valueAsString(l)
-	for _, k := range l.lockedMutexes {
-		if k == s {
-			return s, true
-		}
-	}
-	return s, false
+
+	lockType := l.getLock(s)
+	isLocked := lockType.is(LOCKED) && lockType.isNot(UNLOCKED)
```

Comment:
So we moving from having LOCKED or UNLOCKED states to having: {}, {LOCKED}, {UNLOCKED} and {LOCKED, UNLOCKED}. This is saying that is held is iff the state is {LOCKED}. It is not {LOCKED, UNLOCKED}. This needs to be checked at each location isHeld is called.

Zvonimir, thoughts?

Reply:
I believe the answer really depends on the precision of the analysis. I believe there are in principle two functions

isDefinitelyHeld -- returns true iff the state is {LOCKED}
isMaybeHeld -- returns true iff the state is {LOCKED, UNLOCKED} (or also {LOCKED})

My point is that when we have {LOCKED, UNLOCKED} it can be because the lock can indeed be in two of these states or because of imprecision of the analysis. We could just start with what you suggested.